EventFilter/RPCRawToDigi/plugins/RPCTwinMuxDigiToRaw.cc

```diff
@@ -0,0 +1,182 @@
+#include "EventFilter/RPCRawToDigi/plugins/RPCTwinMuxDigiToRaw.h"
+
+#include <cstring>
+#include <memory>
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/CRC16.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include "CondFormats/DataRecord/interface/RPCInverseLBLinkMapRcd.h"
+#include "CondFormats/DataRecord/interface/RPCInverseTwinMuxLinkMapRcd.h"
+#include "CondFormats/RPCObjects/interface/RPCAMCLink.h"
+#include "CondFormats/RPCObjects/interface/RPCAMCLinkMap.h"
+#include "CondFormats/RPCObjects/interface/RPCInverseAMCLinkMap.h"
+#include "CondFormats/RPCObjects/interface/RPCInverseLBLinkMap.h"
+#include "DataFormats/FEDRawData/interface/FEDHeader.h"
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+#include "DataFormats/FEDRawData/interface/FEDTrailer.h"
+#include "DataFormats/RPCDigi/interface/RPCDigiCollection.h"
+#include "EventFilter/RPCRawToDigi/interface/RPCTwinMuxPacker.h"
+#include "EventFilter/RPCRawToDigi/interface/RPCTwinMuxRecord.h"
+
+RPCTwinMuxDigiToRaw::RPCTwinMuxDigiToRaw(edm::ParameterSet const & _config)
+    : bx_min_(_config.getParameter<int>("bxMin"))
+    , bx_max_(_config.getParameter<int>("bxMax"))
+    , ignore_eod_(_config.getParameter<bool>("ignoreEOD"))
+    , event_type_(_config.getParameter<int>("eventType"))
+    , ufov_(_config.getParameter<unsigned int>("uFOV"))
+{
+    produces<FEDRawDataCollection>();
+    digi_token_ = consumes<RPCDigiCollection>(_config.getParameter<edm::InputTag>("inputTag"));
+}
+
+RPCTwinMuxDigiToRaw::~RPCTwinMuxDigiToRaw()
+{}
+
+void RPCTwinMuxDigiToRaw::fillDescriptions(edm::ConfigurationDescriptions & _descs)
+{
+    edm::ParameterSetDescription _desc;
+    _desc.add<edm::InputTag>("inputTag", edm::InputTag("simMuonRPCDigis", ""));
+    _desc.add<int>("bxMin", -2);
+    _desc.add<int>("bxMax", 2);
+    _desc.add<bool>("ignoreEOD", true);
+    _desc.add<int>("eventType", 1);
+    _desc.add<unsigned int>("uFOV", 1);
+    _descs.add("RPCTwinMuxDigiToRaw", _desc);
+}
+
+void RPCTwinMuxDigiToRaw::beginRun(edm::Run const & _run, edm::EventSetup const & _setup)
+{
+    if (es_tm_link_map_watcher_.check(_setup)) {
+        edm::ESHandle<RPCAMCLinkMap> _es_tm_link_map;
+        _setup.get<RPCTwinMuxLinkMapRcd>().get(_es_tm_link_map);
+        fed_amcs_.clear();
+        for (RPCAMCLinkMap::map_type::const_iterator _tm_link = _es_tm_link_map->getMap().begin()
```

Comment:
looks like here and below many of the for loops can be replaced with much shorter for range loops
```
for (auto const& tm_link : es_tm_link_map->getMap() ){
RPCAMCLink amc(tm_link.first);
... 
}
```